clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java

```diff
@@ -1450,6 +1449,55 @@ public void run() {
     }
   }
 
+  private void waitForSyncOffsetCmd(StoreBufferService.QueueNode node, PubSubTopicPartition topicPartition)
```

Comment:
I think it is better to let `execSyncOffsetCommandAsync` to return a customized `CompletableFuture` and the future.cancel will invalidate the scheduled command.
`StoreBufferService.QueueNode` is more like some implementation detail and normally `SIT` shouldn't use it directly.
WDYT?

Reply:
I think the only difference is that in this implementation, `invalidate` needs to acquire a lock first and this lock enforces  a point that if the `runnable` is already executing, then `invalidate` call has to wait for `runnable` to finish. The is useful as it will prevent the SIT thread from continuing to `close`. Concurrent SIT `close` and `runnable` executing is bad because the internal state of SIT is not reliable anymore  and `runnable` (syncOffset) might not work properly. 

```
    public synchronized void executeSync(Runnable runnable) {
      if (valid) {
        runnable.run();
      }
      cmdExecutedFuture.complete(null);
    }
```
I feel like it is not quite easy to achieve the above if we return a `Future`, but maybe I am wrong here.

Reply:
The Future can be a **customized** `Future` implementation, which can keep a reference to whatever field is required.
To the user of `StoreBufferService`, it shouldn't need to worry about the internal implementation detail, such as `QueueNode`, which we don't expose it from the beginning.

Reply:
I must missed the `customized` keyword in your first comment :-). Yeah, I also don't like the idea to expose the internal `QueueNode` out, but didn't think of an alternative. It's a good idea with a customized future and I will give a try. Thanks for suggestion!